EIPS/eip-FOCIL.md

```diff
@@ -162,11 +161,24 @@ This EIP introduces backward incompatible changes to the block validation rule s
 
 ### IL Equivocation
 
-To mitigate local IL equivocation, FOCIL introduces a new P2P network rule that allows forwarding up to two local ILs per IL committee member. If the proposer or attesters detect two different local ILs sent by the same IL committee member, they should ignore all local ILs from that member. In the worst case, the bandwidth of the local IL gossip subnet can at most double.
+To mitigate IL equivocation, FOCIL introduces a new P2P network rule that allows forwarding up to two ILs per IL committee member. If the proposer or attesters detect two different ILs sent by the same IL committee member, they should ignore all ILs from that member. In the worst case, the bandwidth of the IL gossip subnet can at most double.
+
+### Payload Construction
+
+The block producer, responsible for constructing the execution payload, must ensure that the IL is satisfied. A naive way to do so would be to build an initial payload in whatever way the builder desires, then execute the following algorithm:
+
+1. Sequentially check validity of any yet-to-be-included IL tx against the post-state. If none is found, payload building is over.
+2. If one is found, append it to the end of the payload and update the post-state. Go back to step 1.
+
+The issue with this simple approach is that, given a set of `n` IL transactions, one might end up needing to do `n + (n-1) + (n-2) + ...` validity checks, so `O(n^2)`. For example, the `n`th tx might be valid while all others are not, but its execution sends balance to the sender of the `(n-1)`th tx, making it valid, and in turn, the `(n-1)`th sends balance to the sender of the `(n-2)`th tx, etc.
+
+To efficiently ensure that all valid IL txs have been included in the payload, builders can adopt a simple strategy: prior to building the payload, they store the `nonce` and `balance` of all Externally Owned Accounts (EOAs) involved in IL transactions. As they construct the payload, builders track these EOAs, maintaining and updating each EOA's `nonce` and `balance` whenever changes occur—specifically, when the `nonce` increments (indicating that a transaction from that EOA has been executed) or when the `balance` changes without a `nonce` increment (e.g., after an Account Abstraction (AA) transaction has interacted with that EOA). 
+
+This tracking allows builders to verify IL transaction validity in real-time, enabling them to add transactions sequentially until all remaining transactions are either invalid because the nonce or balance of the associated EOA does not change or cannot be included due to insufficient gas. This approach minimizes overhead by keeping track only of the state changes that are relevant to the validity of IL txs.
```

Comment:
From my understanding, the update of `nonce` and `balance` being tracked also takes O(n). As the update happens each time a IL tx is appended, so the time complexity still O(n^2).

Could you elaborate how does this tracking help block building? Isn't the tracking is to verify if each IL tx is satisfied by the payload easily even with AA transactions when you have transaction summaries but not full transaction data?

Reply:
Perhaps this was meant to mention IL hint?

Reply:
no the time complexity is O(n) overall, because the mechanism ensures each IL txn is validated in constant time relative to the number of transactions:
- Each transaction is validated only once against state.
- Only tracking nonce and balance is more efficient than fully executing txns every time

Reply:
How does the tracking during the payload construction work exactly? I thought it works by querying each EOA's nonce and balance after appending a IL tx.

Reply:
We had a chat about this and finally I'm on the same page. Really appreciate for the explanation @soispoke. Firstly, this is about verifying if the payload satisfies IL constraint by checking if none of not included IL txs is includable. I thought that the tracking method was a solution for the O(n^2) problem in the previous paragraph. After I got off from this misunderstanding, everything got clear.

Anyway, how the verification works is follow:

i) Construct a hashmap where key is `address` and value is a tuple of `nonce` and `balance`. The addresses are the sender of IL txs and `address` field of EIP-7702's new transaction type.
ii) For each remnant IL tx, check if either of `nonce` and `balance` of its relevant `address` is changed by simulating the IL tx. If it does, the payload fails to meet IL constraint.

i) takes O(n) and in ii), each lookup can happen at most twice and there can be at most O(n) remnant IL txs. Thus, the total time complexity is O(n).